Sources/MaliciousSiteProtection/API/APIClient.swift

```diff
@@ -18,129 +18,119 @@
+                ])
+            case .matches(let configuration):
+                endpoint.appendingPathComponent(APIPath.matches).appendingParameter(name: QueryParameter.hashPrefix, value: configuration.hashPrefix)
+            }
+        }
 
-    public init(environment: Environment = .production, session: URLSessionProtocol = URLSession.defaultSession) {
-        switch environment {
-        case .production:
-            endpointURL = Constants.productionEndpoint
-        case .staging:
-            endpointURL = Constants.stagingEndpoint
+        public func headers(for request: APIClient.Request) -> APIRequestV2.HeadersV2 {
+            defaultHeaders
         }
-        self.session = session
     }
 
-    public func getFilterSet(revision: Int) async -> FiltersChangeSetResponse {
-        guard let url = createURL(for: .filterSet, revision: revision) else {
-            logDebug("🔸 Invalid filterSet revision URL: \(revision)")
-            return FiltersChangeSetResponse(insert: [], delete: [], revision: revision, replace: false)
-        }
-        return await fetch(url: url, responseType: FiltersChangeSetResponse.self) ?? FiltersChangeSetResponse(insert: [], delete: [], revision: revision, replace: false)
+}
+
+public struct APIClient: APIClientProtocol {
+
+    let environment: APIClientEnvironment
+    private let service: APIService
+
+    public init(environment: Self.DefaultEnvironment = .production, service: APIService = DefaultAPIService(urlSession: .shared)) {
+        self.init(environment: environment as APIClientEnvironment, service: service)
     }
 
-    public func getHashPrefixes(revision: Int) async -> HashPrefixesChangeSetResponse {
-        guard let url = createURL(for: .hashPrefix, revision: revision) else {
-            logDebug("🔸 Invalid hashPrefix revision URL: \(revision)")
-            return HashPrefixesChangeSetResponse(insert: [], delete: [], revision: revision, replace: false)
-        }
-        return await fetch(url: url, responseType: HashPrefixesChangeSetResponse.self) ?? HashPrefixesChangeSetResponse(insert: [], delete: [], revision: revision, replace: false)
+    public init(environment: APIClientEnvironment, service: APIService) {
+        self.environment = environment
+        self.service = service
     }
 
-    public func getMatches(hashPrefix: String) async -> [Match] {
-        let queryItems = [URLQueryItem(name: "hashPrefix", value: hashPrefix)]
-        guard let url = createURL(for: .matches, queryItems: queryItems) else {
-            logDebug("🔸 Invalid matches URL: \(hashPrefix)")
-            return []
-        }
-        return await fetch(url: url, responseType: MatchResponse.self)?.matches ?? []
+    public func load<Request: APIRequestProtocol>(_ requestConfig: Request) async throws -> Request.ResponseType {
+        let requestType = requestConfig.requestType
+        let headers = environment.headers(for: requestType)
+        let url = environment.url(for: requestType)
```

Comment:
Are there any specific client caching configs we should consider here? The server should return a TTL of 10 minutes for the API so I imagine this will be automatically respected here, but just checking.

Reply:
I‘d say no as the caching policies should be set on the backend side: if there‘s caching it should be fine as all of our requests are parametrized (i.e. revision, hash etc)